docs/architecture/adr-075-rpc-subscription.md

```diff
@@ -0,0 +1,668 @@
+
+The node will retain all events within the time window, provided they do not
+exceed the maximum number.  These config parameters allow the operator to
+loosely regulate how much memory and storage the node allocates to the event
+log.
+
+The event log is shared among all subscribers to the node.
+
+> **Open question:** Should events persist across node restarts?
+>
+> The current event API does not persist events. Some events may be _indexed_,
+> but the node does not keep a complete record: Only BeginBlock and EndBlock
+> events are passed to the indexer, if an indexer is configured at all.
+>
+> The initial implementation will not persist events across restarts.
+
+> **Open question:** The node already stores some event data persistently,
+> notably the events that correspond with block updates. I'm not sure whether
+> it persists other intermediate events, but I think it does not. The event log
+> could be combined with other storage (e.g., reuse the database shared by the
+> blockstore and statestore). Unless we need events to persist, however, it is
+> simpler, easier, and faster to buffer in memory.
+
+### Query API
+
+To retrieve event data, the client will call the (new) RPC method `events`.
+The parameters of this method will correspond to the following Go types:
+
+```go
+type EventParams struct {
+    // Optional filter spec. If nil or empty, all items are eligible.
+    Filter *Filter `json:"filter"`
+
+    // The maximum number of eligible results to return.
+    // If zero or negative, the server will report a default number.
+    MaxResults int `json:"max_results"`
+
+    // Return only items after this cursor. If empty, the limit is just
+    // before the the beginning of the event log.
+    After string `json:"after_item"`
+
+    // Return only items before this cursor.  If empty, the limit is just
+    // after the head of the event log.
+    Before string `json:"before_item"`
+
+    // Wait for up to this long for events to be available.
+    WaitTime time.Duration `json:"wait_time"`
+}
+
+type Filter struct {
+    Query string `json:"query"`
+}
+```
+
+> **Open question:** For the initial implementation, the server will not cache
+> filter queries for the client. If this turns out to be a problem, we can add
+> methods to and and remove stored queries. If necessary, the Query type can be
+> extended to include a query ID option.
+
+> **Open question:** For the initial implementation, I am planning to keep the
+> existing query language as-is. Another option would be to compile queries on
+> the client side, and provide a more structured representation in the request
+> than just a plain string.
+>
+> I think this is probably worth doing, even if only to offload the parsing
+> step onto the client. For now, though, I've stuck to the existing format to
+> make migration easier. The service can keep a small shared cache of compiled
+> queries if performance turns out to be an issue.
```

Comment:
Will it be easy/non-breaking to do this?

Reply:
I'm sorry, which "this" do you mean? Probably "yes" in either case, though.

- If we add a structured representation, it would be in _addition_ to the query string, since we want to continue to support existing queries (which are all strings). For clients in Go, we could transparently compile those strings to the new representation in a library. Other languages would need to write something (but we can support the flat query strings more or less indefinitely).
- Compiling and caching queries server-side is a completely transparent optimization. In fact, we used to do this, but I sped it up so much in #7319 that we didn't need to do that anymore. On that basis, I don't think this is a problem, but we aren't painting ourselves into a corner if I'm wrong.

Reply:
I meant the first of your points. Thanks for the context.